clang/lib/CIR/Dialect/IR/CIRDialect.cpp

```diff
@@ -393,6 +393,14 @@ static LogicalResult checkConstantTypes(mlir::Operation *op, mlir::Type opType,
     return op->emitOpError("zero expects struct or array type");
   }
 
+  if (isa<mlir::cir::UndefAttr>(attrType)) {
+    // Per the LLVM LangRef, "Undefined values may be of any type (other than
+    // 'label' or 'void')". We don't have label types so we just check for void.
```

Comment:
Please remove the reference to LLVM, let's not tie CIR to LLVM constraints - it's possible LLVM might not even be in the picture in a given pipeline.

Is there any C/C++ code that can lead to an undef used with void? I think it's totally fine for composability if CIR has a undef void (if there's C/C++ code that can lead to this). How that behaves during LLVM lowering is a different story.

Reply:
This hasn't been addressed.